classes/plugins/PluginHelper.php

```diff
@@ -192,74 +192,99 @@ protected function _getConnectionParams()
+            }
 
-        // Check whether deleting has worked.
-        if (is_dir($pluginDest)) {
-            throw new Exception(__('manager.plugins.deleteError', ['pluginName' => $pluginVersion->getProduct()]));
-        }
+            // Copy the plug-in from the temporary folder to the target folder.
+            if (!$fileManager->copyDir($sourcePath, $destinyPath)) {
+                throw new Exception('Could not copy plugin to destination!');
+            }
 
-        // Copy the plug-in from the temporary folder to the target folder.
-        if (!$fileManager->copyDir($path, $pluginDest)) {
-            throw new Exception('Could not copy plugin to desination!');
-        }
-        if (!$fileManager->rmtree($path)) {
-            throw new Exception('Could not remove temporary plugin path!');
+            try {
+                $upgradeFile = "{$destinyPath}/" . self::PLUGIN_UPGRADE_FILE;
+                if ($fileManager->fileExists($upgradeFile)) {
+                    $siteDao = DAORegistry::getDAO('SiteDAO'); /** @var SiteDAO $siteDao */
+                    $site = $siteDao->getSite();
+                    $params = $this->_getConnectionParams();
+                    $params['locale'] = $site->getPrimaryLocale();
+                    $params['additionalLocales'] = $site->getSupportedLocales();
+                    $installer = new Upgrade($params, $upgradeFile, true);
+
+                    if (!$installer->execute()) {
+                        throw new Exception(__('manager.plugins.upgradeFailed', ['errorString' => $installer->getErrorString()]));
+                    }
+                }
+
+                $pluginVersion->setCurrent(1);
+                $versionDao->insertVersion($pluginVersion, true);
+                return $pluginVersion;
+            } catch (Throwable $e) {
+                // Delete the plugin files on failure
+                $fileManager->rmtree($destinyPath);
+                throw $e;
+            }
+        } finally {
+            $fileManager->rmtree($sourcePath);
         }
+    }
 
-        $upgradeFile = $pluginDest . '/' . self::PLUGIN_UPGRADE_FILE;
-        if ($fileManager->fileExists($upgradeFile)) {
-            $siteDao = DAORegistry::getDAO('SiteDAO'); /** @var SiteDAO $siteDao */
-            $site = $siteDao->getSite();
-            $params = $this->_getConnectionParams();
-            $params['locale'] = $site->getPrimaryLocale();
-            $params['additionalLocales'] = $site->getSupportedLocales();
-            $installer = new Upgrade($params, $upgradeFile, true);
-
-            if (!$installer->execute()) {
-                throw new Exception(__('manager.plugins.upgradeFailed', ['errorString' => $installer->getErrorString()]));
+    /**
+     * Attempts to create a locked and writable temporary file
+     * The prefix/suffix will receive a minor sanitization
+     */
+    public static function getTemporaryFile(string $prefix = '', string $suffix = ''): SplFileObject
```

Comment:
I was going to move this method to the `FileManager` class, but I wasn't sure if that class is going to be dropped later.

I found a similar method at the Symfony/Filesystem (which is available indirectly), but it returns a resource, which isn't very interesting 🤔

As a second note, I think the Symfony/Filesystem would be a good base for our `FileManager`, it's boxing the I/O operations (if an error happens at the `copy`, it throws an exception).